File: src/main/java/no/fint/provider/springer/model/OkonomiFakturaRepository.java
package no.fint.provider.springer.model;

import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.Event;
import no.fint.event.model.ResponseStatus;
import no.fint.event.model.Status;
import no.novari.fint.model.okonomi.faktura.FakturaActions;
import no.novari.fint.model.resource.FintLinks;
import no.novari.fint.model.resource.okonomi.faktura.FakturautstederResource;
import no.fint.provider.springer.storage.SpringerRepository;
import org.springframework.stereotype.Repository;

import java.util.Collections;
import java.util.Set;

@Slf4j
@Repository
public class OkonomiFakturaRepository extends SpringerRepository {

    @Override
    public void accept(Event<FintLinks> response) {
        switch (FakturaActions.valueOf(response.getAction())) {
            case GET_ALL_FAKTURAUTSTEDER:
                query(FakturautstederResource.class, response);
                break;
            default:
                response.setStatus(Status.ADAPTER_REJECTED);
                response.setResponseStatus(ResponseStatus.REJECTED);
                response.setStatusCode("INVALID_ACTION");
                response.setMessage("Invalid action");
        }
    }

    @Override
    public Set<String> actions() {
        return Collections.singleton(FakturaActions.GET_ALL_FAKTURAUTSTEDER.name());
    }
}

File: src/main/kotlin/no/fint/provider/springer/model/OkonomiFakturaRepository.kt
package no.fint.provider.springer.model

import no.fint.event.model.Event
import no.fint.event.model.ResponseStatus
import no.fint.event.model.Status
import no.fint.provider.springer.storage.SpringerRepository
import no.fint.provider.springer.storage.Wrapper
import no.novari.fint.model.okonomi.faktura.FakturaActions
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.resource.okonomi.faktura.FakturautstederResource
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Repository
import java.util.stream.Collectors
import java.util.stream.Stream

@Repository
class OkonomiFakturaRepository(
    mongoTemplate: MongoTemplate,
    wrapper: Wrapper
) : SpringerRepository(mongoTemplate, wrapper) {

    override fun accept(response: Event<FintLinks>) {
        when (FakturaActions.valueOf(response.action)) {
            FakturaActions.GET_ALL_FAKTURAUTSTEDER -> query(FakturautstederResource::class.java, response)

            else -> invalid(response)
        }
    }

    private fun invalid(response: Event<FintLinks>) {
        response.status = Status.ADAPTER_REJECTED
        response.responseStatus = ResponseStatus.REJECTED
        response.statusCode = "INVALID_ACTION"
        response.message = "Invalid action"
    }

    override fun actions(): Set<String> =
        Stream.of(*FakturaActions.values())
            .map { it.name }
            .filter { it.startsWith("GET_ALL_") }
            .collect(Collectors.toSet())
}
